neural_network_lyapunov/relu_to_optimization.py

```diff
@@ -561,6 +562,50 @@ def _compute_layer_bound(self,
                             x_lo.detach().numpy(),
                             x_up.detach().numpy(),
                             create_prog_callback, binary_var_type)
+                if method == mip_utils.PropagateBoundsMethod.IA_MIP:
+                    # We also compute the bounds by IA. If the IA bound is
+                    # close to the MIP bound, then we use the IA bound.
+                    # Otherwise, we relax the MIP bound a little bit to make
+                    # sure this relaxed bound will not be active.
+                    z_pre_relu_lo_ia, z_pre_relu_up_ia = \
+                        mip_utils.compute_range_by_IA(
+                            self.model[2 * layer_count].weight, bias,
+                            linear_layer_input_lo, linear_layer_input_up)
+                    for j in range(self.model[2 * layer_count].out_features):
+                        neuron_index = self.relu_unit_index[layer_count][j]
+                        if torch.abs(z_pre_relu_lo[neuron_index] -
+                                     z_pre_relu_lo_ia[j]) < 1E-5:
+                            z_pre_relu_lo[neuron_index] = z_pre_relu_lo_ia[j]
+                        else:
+                            if z_pre_relu_lo[neuron_index] > 0:
+                                # If the MIP lower bound is > 0, then the ReLU
+                                # unit should always be active. We reduce this
+                                # MIP lower bound a little bit (but still
+                                # remains positive)
+                                z_pre_relu_lo[neuron_index] *= 0.99
+                                z_pre_relu_lo[
+                                    neuron_index] += 0.01 * torch.maximum(
```

Comment:
maybe just use torch.clamp for this? avoids creating a tensor here (and having to go back to add .device on everything once we move to GPUs ;) )

Reply:
Done. Good call.